businessRules/donations.js

```diff
@@ -1,54 +1,59 @@
 const { v4 } = require('uuid');
 const { createDonation } = require('../models/donations');
-const { getAllIdeas } = require('../models/ideas');
-const { getIdeas } = require('../utils/storage');
+const { validateIdea } = require('../validations/ideas');
+const { validateSum } = require('../validations/donations.js');
+const { getUser } = require('./users');
+const { getIdea } = require('./ideas');
 
-function createDonation(firstName, sum, userId, ideaId) {
-    let ideas = getIdeas();
-    const idea = ideas.find((idea) => ideaId === idea.id);
+function createAnonymousUser(firstName, userId) {
+    userId = v4();
+    const anonymous = 'Anonymous';
+
+    firstName = firstName || anonymous;
+}
 
-    if (!idea) {
-        throw new Error('Error. No idea with such ID found.');
+function createDonationByUnregisteredUser(sum, firstName, ideaId) {
+    validateIdea(ideaId);
+    validateSum(sum);
+    let userId;
+    if (!firstName) {
+        createAnonymousUser(firstName, userId);
     }
+    createDonation(firstName, sum, userId, ideaId);
+}
 
-    const generateId = v4();
-    const anonymous = 'Anonymous';
+function createDonationByRegisteredUser(sum, userId, ideaId) {
+    validateSum(sum);
 
-    firstName = firstName || anonymous;
-    userId = userId || generateId;
+    const user = getUser(userId);
+
+    const firstName = user.firstName;
 
     createDonation(firstName, sum, userId, ideaId);
 }
 
-function ideasDonationSum(ideaId) {
-    const ideas = getAllIdeas();
-    const idea = ideas.find((idea) => idea.id === ideaId);
+function getTotalSumDonatedForIdea(ideaId) {
+    const idea = getIdea(ideaId);
 
-    if (!idea) {
-        throw new Error('Error. No idea with such ID found.');
-    }
     return idea.totalDonationSum;
 }
 
-function ideasSumDifference(ideaId) {
-    const ideas = getAllIdeas();
-    const idea = ideas.find((idea) => idea.id === ideaId);
-
-    if (!idea) {
-        throw new Error('Error. No idea with such ID found.');
-    }
+function getIdeasSumLeftToDonate(ideaId) {
```

Comment:
Why not do it the other way around: create a function that calculates how much was already donated?

Reply:
But I have one:
getTotalSumDonatedForIdea
I've decided to calculate how much left.